sdk/servicebus/service-bus/src/track2/serviceBusReceiverClient.ts

```diff
@@ -0,0 +1,526 @@
+   * @param options Options for receiveBatch.
+   */
+  receiveBatch(
+    maxMessages: number,
+    maxWaitTimeInSeconds?: number,
+    options?: ReceiveBatchOptions
+  ): Promise<Message[]>;
+
+  /**
+   * Closes the client.
+   */
+  close(): Promise<void>;
+
+  /**
+   * Methods related to service bus diagnostics.
+   */
+  diagnostics: {
+    /**
+     * Peek within a queue or subscription.
+     * NOTE: this method does not respect message locks or increment delivery count
+     * for messages.
+     * @param maxMessageCount The maximum number of messages to retrieve.
+     */
+    peek(maxMessageCount?: number): Promise<Message[]>;
+
+    /**
+     * Peek within a queue or subscription, starting with a specific sequence number.
+     * NOTE: this method does not respect message locks or increment delivery count
+     * for messages.
+     * @param fromSequenceNumber The sequence number to start peeking from (inclusive).
+     * @param maxMessageCount The maximum number of messages to retrieve.
+     */
+    peekBySequenceNumber(fromSequenceNumber: Long, maxMessageCount?: number): Promise<Message[]>;
+  };
+}
+
+/**
+ * Methods to manage rules for subscriptions. More information about subscription rules
+ * can be found here: https://docs.microsoft.com/en-us/azure/service-bus-messaging/topic-filters
+ */
+export interface SubscriptionRuleManagement {
+  getRules(): Promise<RuleDescription[]>;
+  removeRule(ruleName: string): Promise<void>;
+  addRule(
+    ruleName: string,
+    filter: boolean | string | CorrelationFilter,
+    sqlRuleActionExpression?: string
+  ): Promise<void>;
+}
+
+export type ClientTypeT<
+  ReceiveModeT extends "peekLock" | "receiveAndDelete",
+  EntityTypeT extends "queue" | "subscription",
+  SessionsEnabledT extends "sessions" | "nosessions"
+> = SessionsEnabledT extends "nosessions"
+  ? EntityTypeT extends "queue"
+    ? NonSessionReceiver<ReceiveModeT>
+    : NonSessionReceiver<ReceiveModeT> & SubscriptionRuleManagement
+  : EntityTypeT extends "queue"
+  ? SessionReceiver<ReceiveModeT>
+  : SessionReceiver<ReceiveModeT> & SubscriptionRuleManagement;
+
+export interface ServiceBusReceiverClient {
+  /**
+   * Creates a client for an Azure Service Bus queue.
+   *
+   * @param queueAuth Data needed to connect to a queue.
+   * @param receiveMode The receive mode to use (defaults to PeekLock)
+   * @param options Options for the client itself.
+   */
+  new (queueAuth: QueueAuth, receiveMode: "peekLock"): ClientTypeT<
+    "peekLock",
+    "queue",
+    "nosessions"
+  >;
+
+  /**
+   * Creates a client for an Azure Service Bus queue.
+   *
+   * @param queueAuth Data needed to connect to a queue.
+   * @param receiveMode The receive mode to use (defaults to PeekLock)
+   * @param options Options for the client itself.
+   */
+  new (queueAuth: QueueAuth, receiveMode: "receiveAndDelete"): ClientTypeT<
+    "receiveAndDelete",
+    "queue",
+    "nosessions"
+  >;
```

Comment:
I think you can condense these into a single overload by doing this:
```suggestion
  new (queueAuth: QueueAuth, receiveMode: "peekLock" | "receiveAndDelete"): ClientTypeT<
    typeof receiveMode,
    "queue",
    "nosessions"
  >;
```

The same strategy can be used for the Session/subscriptionAuth versions as well. I thought that might make documenting the `receiveMode` easier but if what you have results in better documentation then disregard this.

Reply:
Nevermind, I was wrong on this one :)